### ISSUE TITLE
`zetacore` : Delete Ballots after distribution

### ISSUE THREAD

**Issue by kingpinXD** [zeta-chain/node#942]
Ballots can be deleted from the state after they have matured to save space on the state.db

- Delete matured ballots
- Add an event when the ballot gets deleted, which can be indexed and used to query the list of voters if needed for debugging

**Comment by ws4charlie** [comment]
Maybe we should not delete the ballots as the finalized ballots help us prevent double spending (by blocking the same ballot being created again)

**Comment by kingpinXD** [comment]
We do have the finalized inbounds check to stop ballot creation 
https://github.com/zeta-chain/zeta-node/blob/db2c6f3ff7998c1f5f8500e853500a5d9d46adf9/x/crosschain/keeper/msg_server_vote_inbound_tx.go#L95-L97

**Comment by ws4charlie** [comment]
> We do have the finalized inbounds check to stop ballot creation https://github.com/zeta-chain/zeta-node/blob/db2c6f3ff7998c1f5f8500e853500a5d9d46adf9/x/crosschain/keeper/msg_server_vote_inbound_tx.go#L95-L97

Sounds great. It will be safe then. We can go ahead to clean up ballots after distribution.

**Comment by kingpinXD** [comment]
Deleting older ballots would remove the possibility of querying ballots older than 100 blocks (current maturity period ) 

This can affect current way of operations 
1. We wouldn't be able to query older ballots to see if directly if a particular observer is consistently missing votes 
2. We allow observers to vote on ballots to receive rewards , they would not be able to vote on ballots once deleted 



Proposed solutions 
When deleting a ballot, we can emit a new event containing the list of observers who have added their votes to the ballot. 
This should be enough, as we expect every observer to cast their vote within the maturity period (casting after that is useless as they would not get any rewards )

- The delete ballot event would be part of the BlockBeginner and not a tx_result.

- We can also consider removing the `Ballot_Created` even if it is not used by any consumer.


Changes required 
- The explorer/indexer should be able to index these deleted ballot events so that they can be easily searched. This should be very similar to existing emissions events such as 
`/zetachain.zetacore.emissions.internal.ObserverEmissions`
`/zetachain.zetacore.emissions.internal.BlockEmissions`


cc @zeta-chain/fullstack , @zeta-chain/devops

**Comment by GMaiolo** [comment]
If these changes come within `/cosmos/tx/v1beta1/txs` when querying for a specific block event, then we don't need to update anything on the indexer side 👍

**Comment by kingpinXD** [comment]
I don't think it would be within `/cosmos/tx/v1beta1/txs`  , as its not part of a tx but part of begin block instead . (Not 100% sure though ) 

Can you send me an example of how we index the two events I mentioned?

**Comment by GMaiolo** [comment]
If it's not in `/cosmos/tx/v1beta1/txs` and if it's not EVM (therefore not going through Blockscout) we're not indexing it currently

**Comment by kingpinXD** [comment]
@GMaiolo  We should be indexing `BeginBlock` and `EndBlock` events as well. In case we are not doing it now , it would be a good idea to prioritize this shortly

**Comment by lumtis** [comment]
https://github.com/zeta-chain/node/pull/2863

**Comment by kingpinXD** [comment]
> Deleting older ballots would remove the possibility of querying ballots older than 100 blocks (current athens maturity period )
> 
> This can affect current way of operations
> 
> 1. We wouldn't be able to query older ballots to see if directly if a particular observer is consistently missing votes
> 2. We allow observers to vote on ballots to receive rewards , they would not be able to vote on ballots once deleted
> 
> Proposed solutions When deleting a ballot, we can emit a new event containing the list of observers who have added their votes to the ballot. This should be enough, as we expect every observer to cast their vote within the maturity period (casting after that is useless as they would not get any rewards ) Additionally, we can consider increasing the maturity blocks, if we expect observers to be more than 100 blocks late in adding their votes
> 
> * The delete ballot event would be part of the BlockBeginner and not a tx_result.
> * We can also consider removing the `Ballot_Created` if it is not being consumed right now
> 
> Changes required
> 
> * The explorer/indexer should be able to index these deleted ballot events so that they can be easily searched. This should be very similar to existing emissions events such as
>   `/zetachain.zetacore.emissions.internal.ObserverEmissions`
>   `/zetachain.zetacore.emissions.internal.BlockEmissions`
> 
> cc @zeta-chain/fullstack , @zeta-chain/devops


The delete ballot event is not mandatory , but we can add it if needed 
However if someone can confirm that it is not needed at all , and that we are good with the ballot stats for the last 100 blocks , I would like to remove it to save space in the data directory 

@GMaiolo @lumtis
